--- ../cramfs-1.1.old/cramfsck.c	2007-02-23 17:44:55.000000000 +0100
+++ ./cramfsck.c	2007-02-23 18:18:42.000000000 +0100
@@ -77,6 +77,8 @@
 #ifdef INCLUDE_FS_TESTS
 static int opt_extract = 0;		/* extract cramfs (-x) */
 static char *extract_dir = "root";	/* extraction directory (-x) */
+static int opt_stdout = 0;		/* extract one cramfs file to stdout (-s) */
+static char *extract_file = "";		/* file to extract (-s) */
 static uid_t euid;			/* effective UID */
 
 /* (cramfs_super + start) <= start_dir < end_dir <= start_data <= end_data */
@@ -107,9 +109,10 @@
 {
 	FILE *stream = status ? stderr : stdout;
 
-	fprintf(stream, "usage: %s [-hv] [-x dir] file\n"
+	fprintf(stream, "usage: %s [-hv] [-x dir] [-s file] file\n"
 		" -h         print this help\n"
 		" -x dir     extract into dir\n"
+		" -s file    extract file to stdout\n"
 		" -v         be more verbose\n"
 		" file       file to test\n", progname);
 
@@ -398,7 +401,7 @@
 			}
 		}
 		size -= out;
-		if (opt_extract) {
+		if (fd >= 0) {
 			if (write(fd, outbuffer, out) < 0) {
 				die(FSCK_ERROR, 1, "write failed: %s", path);
 			}
@@ -495,7 +498,7 @@
 static void do_file(char *path, struct cramfs_inode *i)
 {
 	unsigned long offset = i->offset << 2;
-	int fd = 0;
+	int fd = -1;
 
 	if (offset == 0 && i->size != 0) {
 		die(FSCK_UNCORRECTED, 0, "file inode has zero offset and non-zero size");
@@ -509,8 +512,11 @@
 	if (opt_verbose) {
 		print_node('f', i, path);
 	}
-	if (opt_extract) {
-		fd = open(path, O_WRONLY | O_CREAT | O_TRUNC, i->mode);
+	if (opt_extract || (opt_stdout && !strcmp(path+strlen(extract_dir), extract_file))) {
+		if (opt_stdout)
+			fd = fileno(stdout);
+		else
+			fd = open(path, O_WRONLY | O_CREAT | O_TRUNC, i->mode);
 		if (fd < 0) {
 			die(FSCK_ERROR, 1, "open failed: %s", path);
 		}
@@ -566,6 +572,11 @@
 		}
 		change_file_status(path, i);
 	}
+	if (opt_stdout && !strcmp(path+strlen(extract_dir), extract_file)) {
+			/* This is a hack */
+			printf("%s\n", outbuffer);
+			exit(FSCK_USAGE);
+	}
 }
 
 static void do_special_inode(char *path, struct cramfs_inode *i)
@@ -674,7 +685,7 @@
 		die(FSCK_ERROR, 1, "failed to allocate outbuffer");
 
 	/* command line options */
-	while ((c = getopt(argc, argv, "hx:v")) != EOF) {
+	while ((c = getopt(argc, argv, "hx:s:v")) != EOF) {
 		switch (c) {
 		case 'h':
 			usage(FSCK_OK);
@@ -683,6 +694,10 @@
 			opt_extract = 1;
 			extract_dir = optarg;
 			break;
+		case 's':
+			opt_stdout = 1;
+			extract_file = optarg;
+			break;
 #else /* not INCLUDE_FS_TESTS */
 			die(FSCK_USAGE, 0, "compiled without -x support");
 #endif /* not INCLUDE_FS_TESTS */
